### PR TITLE
GPU ams ame ads

### PR DESCRIPTION
**Pull request by liruipeng**
This PR adds GPU support for ams, ame and ads, 
- [x] ams
- [x] ame
- [x] ads (with no testings)

for which some ParCSR modules are also added on GPUs

1. ParCSRAdd
2. ParCSRTranspose
3. l1 hybrid G-S/SSOR

Regression tests

- [x] tux
- [x] ray
- [x] lassen

### CODE REVIEW

**Review by AerysL — COMMENTED** [review]
Looked over, made a few comments. Don't see anything that I think that is a major block

File: src/parcsr_ls/ame.c [hunk 1]
```diff
@@ -271,27 +349,56 @@ HYPRE_Int hypre_AMESetup(void *esolver)
...
+                               edge_bc + ne,
+                               l1norm_arr,
+                               less_than<HYPRE_Real>(eps),
+                               1 );
+            hypre_TFree(l1norm_arr, memory_location);
+         }
+         else
+#endif
+         {
+            HYPRE_Real l1_norm;
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/parcsr_ls/ame.c:384]: @AerysL Technically we can, but we don't currently have an option to exclude diagonal entries in `Ad`. I could do the same thing as in the GPU path, but I was trying to not change the CPU code in `ams.c`.

File: src/parcsr_ls/ams.c [hunk 2]
```diff
@@ -244,7 +302,7 @@ HYPRE_Int hypre_BoomerAMGBlockSolve(void *B,
  * For every zero row in the matrix: set the diagonal element to 1.
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/parcsr_ls/ams.c:306]: It computes l1 norms and fix zero rows based on that. We also have routine computing l1-norms and return them. This one is kinda special for ams.

File: src/parcsr_mv/par_csr_matop.c [hunk 3]
```diff
@@ -5118,11 +5147,6 @@ hypre_ParCSRMatrixAdd( HYPRE_Complex        alpha,
 
    HYPRE_ANNOTATE_FUNC_BEGIN;
 
-   hypre_assert(num_rows_A == num_rows_B);
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/parcsr_mv/par_csr_matop.c:5121]: These were moved to `hypre_ParCSRMatrixAdd`

File: src/examples/ex15.c [hunk 4]
```diff
@@ -767,7 +767,7 @@ int main (int argc, char *argv[])
       int part = 0;
       int var = 0; /* the node variable */
       int index[3];
-      double xval, yval, zval;
+      double *xyzval = (double *) malloc(3*sizeof(double));
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/examples/ex15.c:770]: @AerysL You meant `xyzval[3]`? Note that we redefine `malloc` to `cudaMallocManaged` for GPU runs.

File: src/test/TEST_examples/gpu.jobs [hunk 5]
```diff
@@ -30,7 +32,10 @@ mpirun -np 16 ./ex9 -n 33 -solver 0 -v 1 1 > gpu.out.9
 
 mpirun -np 2 ./ex12 -pfmg > gpu.out.12
 
+# mpirun -np 2 ./ex12f -pfmg > gpu.out.12f
+
 mpirun -np 6 ./ex13 -n 10 > gpu.out.13
 
 mpirun -np 6 ./ex14 -n 10 > gpu.out.14
 
+# mpirun -np 8 ./ex15 -n 10 > gpu.out.15
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/test/TEST_examples/gpu.jobs:41]: This example has some problems and is extremely slow on GPU with UVM. Comment it out for now

File: src/test/sstruct.c [hunk 6]
```diff
@@ -3357,7 +3363,7 @@ main( hypre_int argc,
                        index[0] <= pdata.iuppers[box][0]; index[0]++)
                   {
                      HYPRE_SStructVectorAddFEMValues(b, part, index,
```
[the remark this answers is not in the document]
Comment (liruipeng) [reply src/test/sstruct.c:3365]: For CPU build yes, since `HYPRE_MEMORY_DEVICE` == HOST. For GPU build, it works with UVM and EXEC_HOST

**Review by AerysL — APPROVED** [review]

File: src/test/runtest.sh [hunk 7]
```diff
@@ -117,10 +117,12 @@ function MpirunString
          ;;
       ray*)
          shift
-         RunString="lrun -n$*"
+         RunString="mpirun -n $*"
```
Comment (rfalgout) [inline src/test/runtest.sh:120]: Why mpirun here and lrun on lassen?

**Review by rfalgout — APPROVED** [review]
Looks good!  I didn't realize it would take quite so much code to port these solvers, so thanks!

File: src/test/runtest.sh (hunk 7 above)
Comment (liruipeng) [reply src/test/runtest.sh:120]: I would like to use `lrun` on ray too, but it has some weird MPI errors (which has been reported to LC). Maybe will change it back when it is fixed.

Comment (rfalgout) [reply src/test/runtest.sh:120]: I see.  This makes sense.  Thanks again!

**Comment by liruipeng** [comment]
> Looks good! I didn't realize it would take quite so much code to port these solvers, so thanks!

@rfalgout It's mainly because some ParCSR building blocks were not there.

File: src/test/TEST_ams/solvers.saved.lassen [hunk 8]
```diff
@@ -1,117 +1,117 @@
...
-    Cycle 41   1.793733e-05    0.722065     1.335871e-06 
-    Cycle 42   1.295116e-05    0.722022     9.645285e-07 
+    Cycle 45   2.495866e-05    0.756777     1.858779e-06 
+    Cycle 46   1.889154e-05    0.756913     1.406935e-06 
+    Cycle 47   1.430435e-05    0.757183     1.065307e-06 
+    Cycle 48   1.083145e-05    0.757214     8.066648e-07 
 
 
- Average Convergence Factor = 0.719067
+ Average Convergence Factor = 0.746545
```
Comment (ulrikeyang) [inline src/test/TEST_ams/solvers.saved.lassen:35]: It looks like problem 2 and 3 converge quite a bit slower now. Are those the same problems as before?

**Review by ulrikeyang — APPROVED** [review]
Overall, this looks fine, but I had a question about the change in convergence for AMS solvers 2 and 3 problem.

File: src/test/TEST_ams/solvers.saved.lassen (hunk 8 above)
Comment (liruipeng) [reply src/test/TEST_ams/solvers.saved.lassen:35]: Same problems. I think the difference was from using PMIS coarsening on GPUs vs. HMIS on CPU.

**Comment by ulrikeyang** [comment]
Ok. That makes sense.

**Review by AerysL — APPROVED** [review]

**Review by waynemitchell — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**